libbeat/reader/multiline/while.go

```diff
@@ -0,0 +1,226 @@
+	}
+
+	matcherFunc := lineMatcher(*config.Pattern)
+	if config.Negate {
+		matcherFunc = negatedLineMatcher(matcherFunc)
+	}
+
+	pr := &whilePatternReader{
+		reader:    r,
+		matcher:   matcherFunc,
+		msgBuffer: newMessageBuffer(maxBytes, maxLines, []byte(separator), config.SkipNewLine),
+		logger:    logp.NewLogger("reader_multiline"),
+		state:     (*whilePatternReader).readFirst,
+	}
+	return pr, nil
+}
+
+// Next returns next multi-line event.
+func (pr *whilePatternReader) Next() (reader.Message, error) {
+	return pr.state(pr)
+}
+
+func (pr *whilePatternReader) readFirst() (reader.Message, error) {
+	for {
+		message, err := pr.reader.Next()
+		if err != nil {
+			// no lines buffered -> ignore timeout
+			if err == sigMultilineTimeout {
+				continue
+			}
+
+			pr.logger.Debug("Multiline event flushed because timeout reached.")
+
+			// pass error to caller (next layer) for handling
+			return message, err
+		}
+
+		if message.Bytes == 0 {
+			continue
+		}
+
+		// no match, return message
+		if !pr.matcher(message.Content) {
+			return message, nil
+		}
+
+		// Start new multiline event
+		pr.msgBuffer.startNewMessage(message)
+		pr.setState((*whilePatternReader).readNext)
+		return pr.readNext()
+	}
+}
+
+func (pr *whilePatternReader) readNext() (reader.Message, error) {
+	for {
+		message, err := pr.reader.Next()
+		if err != nil {
+			// handle multiline timeout signal
+			if err == sigMultilineTimeout {
+				// no lines buffered -> ignore timeout
+				if pr.msgBuffer.isEmpty() {
+					continue
+				}
+
+				pr.logger.Debug("Multiline event flushed because timeout reached.")
+
+				// return collected multiline event and
+				// empty buffer for new multiline event
+				msg := pr.msgBuffer.finalize()
+				pr.resetState()
+				return msg, nil
+			}
+
+			// handle error without any bytes returned from reader
+			if message.Bytes == 0 {
+				// no lines buffered -> return error
+				if pr.msgBuffer.isEmpty() {
+					return reader.Message{}, err
+				}
+
+				// lines buffered, return multiline and error on next read
+				return pr.collectMessageAfterError(err)
+			}
+
+			// handle error with some content being returned by reader and
+			// line matching multiline criteria or no multiline started yet
+			if pr.msgBuffer.isEmptyMessage() || pr.matcher(message.Content) {
+				pr.msgBuffer.addLine(message)
+
+				// return multiline and error on next read
+				return pr.collectMessageAfterError(err)
+			}
+
+			// no match, return current multiline and return current line on next
+			// call to readNext
+			msg := pr.msgBuffer.finalize()
+			pr.msgBuffer.load(message)
+			pr.setState((*whilePatternReader).notMatchedMessageLoad)
+			return msg, nil
+		}
+
+		// no match, return message if buffer is empty, otherwise return current
+		// multiline and save message to buffer
+		if !pr.matcher(message.Content) {
+			if pr.msgBuffer.isEmptyMessage() {
+				return message, nil
+			} else {
```

Comment:
Nit: It looks better if you do not use an `else` branch.

```golang
if pr.msgBuffer.isEmptyMessage() {
	return message, nil
}
msg := pr.msgBuffer.finalize()
pr.msgBuffer.load(message)
pr.setState((*whilePatternReader).notMatchedMessageLoad)
return msg, nil
```

Reply:
ok